hare/algorithm.go

```diff
@@ -0,0 +1,230 @@
+package hare
+
+import (
+	"errors"
+	"github.com/gogo/protobuf/proto"
+	"github.com/spacemeshos/go-spacemesh/crypto"
+	"github.com/spacemeshos/go-spacemesh/hare/pb"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+	"sync"
+	"time"
+)
+
+const ProtoName = "HARE_PROTOCOL"
+const RoundDuration = time.Second * time.Duration(15)
+const InitialKnowledgeSize = 1000
+const f = 100
+
+type Byteable interface {
+	Bytes() []byte
+}
+
+type NetworkService interface {
+	RegisterProtocol(protocol string) chan service.Message
+	Broadcast(protocol string, payload []byte) error
+}
+
+type State struct {
+	k    uint32      // the iteration number
+	ki   uint32      // indicates when S was first committed upon
+	s    Set         // the set of blocks
+	cert Certificate // the certificate
+}
+
+type ConsensusProcess struct {
+	State
+	Closer // the consensus is closeable
+	pubKey      crypto.PublicKey
+	layerId     LayerId
+	oracle      Rolacle // roles oracle
+	signing     Signing
+	network     NetworkService
+	startTime   time.Time // TODO: needed?
+	inbox       chan *pb.HareMessage
+	knowledge   []*pb.HareMessage
+	stateMutex  sync.Mutex
+	roundMsg    []byte
+	// TODO: add knowledge of notify messages. What is the life-cycle of such messages? persistent according to Julian
+}
+
+func NewConsensusProcess(key crypto.PublicKey, layer LayerId, s Set, oracle Rolacle, signing Signing, p2p NetworkService) *ConsensusProcess {
+	proc := &ConsensusProcess{}
+	proc.State = State{0, 0, s, Certificate{}}
+	proc.Closer = NewCloser()
+	proc.pubKey = key
+	proc.layerId = layer
+	proc.oracle = oracle
+	proc.signing = signing
+	proc.network = p2p
+	proc.knowledge = make([]*pb.HareMessage, 0, InitialKnowledgeSize)
+	m, err := proc.buildStatusMessage()
+	if err != nil {
+		log.Error("could not build status message")
+		proc.roundMsg = nil
+	} else {
+		proc.roundMsg = m
+	}
+
+	return proc
+}
+
+func (proc *ConsensusProcess) Start() error {
+	if !proc.startTime.IsZero() { // called twice on same instance
+		log.Error("ConsensusProcess has already been started.")
+		return StartInstanceError(errors.New("instance already started"))
+	}
+
+	proc.startTime = time.Now()
+
+	go proc.eventLoop()
+
+	return nil
+}
+
+func (proc *ConsensusProcess) createInbox(size uint32) chan *pb.HareMessage {
+	proc.inbox = make(chan *pb.HareMessage, size)
+	return proc.inbox
+}
+
+func (proc *ConsensusProcess) eventLoop() {
+	log.Info("Start listening")
+
+	ticker := time.NewTicker(RoundDuration)
+
+	for {
+		select {
+		case msg := <-proc.inbox: // msg event
+			proc.handleMessage(msg)
+		case <-ticker.C: // next round event
+			proc.nextRound()
+		case <-proc.CloseChannel(): // close event
+			log.Info("Stop event loop, instance aborted")
+			return
+		}
+	}
+}
+
+func (proc *ConsensusProcess) handleMessage(m *pb.HareMessage) {
+	// Note: layer is already verified by the broker
+
+	// verify iteration
+	if m.Message.K != proc.k {
```

Comment:
add a log print

Reply:
Also, add to the comment above that currently we are rejecting past/future messages

Reply:
Fixed